README.md
# REDCap First Responder COVID-19 ETL Engine

This project provides extract, transform, and load (ETL) tools in support of the First Responders COVID-19 Testing project at the University of Florida. The ETL tools are RScripts run by a Docker container.

## Prerequisites

This script use R and these R packages:

    tidyverse
    dotenv
    REDCapR
    openxlsx
    sendmailR

To build the Docker container, you will need only Docker.

This project is designed to move data between two REDCap projects that work together to collect and curate the data in a COVID-19 testing workflow. The source project, referred to as the _results_ project in the code and configuration files, is provided as a REDCap project XML file at [`./examples/First_Responder_COVID19_Results_Upload.xml`](./examples/First_Responder_COVID19_Results_Upload.xml).  The target project, referred to as the _survey_ project in the code and configuration files, is available at as [`First_Responder_COVID19.xml`](https://github.com/ctsit/fr_covidata/blob/master/example/First_Responder_COVID19.xml) in the [fr_covidata REDCap module](https://github.com/ctsit/fr_covidata).

This script uses the REDCap API to move data between the two projects. The API must be enabled on the REDCap project and the host where this script runs will need to have access to it.


## Setup and Configuration

This script is configured entirely via the environment. An example `.env` files are provided as `./example.env` and `./example_pky.env` To use one of these files, copy it to the name `.env` and customize according to your project needs. Follow these steps to build the required components and configure the script's `.env` file.

1. Create each of the REDCap projects from the project XML files. We will refer to these two projects as _survey_ and _results_ for the remainder of this document.
1. In both the survey and results projects, give a user User Rights of _Full Data Set_ for _Data Exports_
1. In each project, that user will need an API key in _each_ project.
1. Add the new API keys to the .env file taking care to not confuse the two keys.
1. Change the `*_PROJECT_TITLE` and `*_PROJECT_PID` fields to match the result and survey projects.
1. Set `TIME_ZONE` to assure that time stamps used in the file names and the email are accurate.
1. Revise the `EMAIL_*` and `SMTP_SERVER` settings to reflect your local needs.


## Running the ETL scripts

Each ETL job in this system is an Rscript run via Docker. The Docker containers are hosted on a Linux host with access to the REDCap API interface and a mail server. At UF, this host is `tools4.ctsi.ufl.edu`. Scripts can also be run at the command line or in RStudio. In each case the script will read its configuration from the `.env` file.

To build the image and run the report using docker within the project directory do:

`docker build -t fr_covidata_engine .`

and run the script using docker with a command something like this:

`docker run --rm --env-file <path_to_dir_full_of_env_files>/fr_dev.env fr_covidata_engine Rscript load_results_into_survey_project.R`

Example cron scripts that could run the containers on a regular basis are provided in [./examples/*.cron](./examples/)


## Release and Deployment

This project uses the Git Flow workflow for releases. Every release should be versioned and have a ChangeLog entry that describes the new features and bug fixes. Every release should also be accompanied by an updated `VERSION` file to allow image builds to be tagged as they are built by the `build.sh`

To deploy a new release on tools4, execute this series of commands or an equivalent from your home directory:

```
git clone https://github.com/ctsit/fr_covidata_engine.git
cd fr_covidata_engine
git pull
sudo ./build.sh
```


## Testing workflows

To test the the `load_results*` scripts, follow these steps:

1. Write the environment file as described above.
1. Create appointment records in the Survey project. 10-15 records make a good test. 
1. Use `load_fake_data_into_upload.R` to generate fake result data in the Results project. It will derive a swab result value from the the research_encounter_id. 
1. Run `load_results_into_pky_projects.R` or `load_results_into_survey_project.R` according to your need.
1. If you are testing the PKY projects with `load_results_into_pky_projects.R`, continue the test by adding appointment records on the next _follow-up_ event in the serial project. 
1. Create fake result data based on these records by running `load_fake_data_from_serial_into_upload.R`.
1. Test the complete workflow by rerunning `load_results_into_pky_projects.R`.


CHANGELOG.md
# Change Log
All notable changes to the REDCap First Responder COVID-19 ETL Engine project will be documented in this file.
This project adheres to [Semantic Versioning](http://semver.org/).


## [0.5.1] - 2020-07-02
### Changed
- Allow indeterminate lab results to be loaded for FR IgG (Philip Chase)
- Remove leading zero from minute in cron file (Philip Chase)


## [0.5.0] - 2020-06-16
### Added
- Add Docker image build script (Philip Chase)


## [0.4.0] - 2020-06-16
### Added
- Add FR IgG load script (Philip Chase)
- Add VERSION file referenced in 'Release and Deployment' (Philip Chase)
- Add Release and Deployment section to README (Philip Chase)
- Add prod cron file for FR (Philip Chase)
- Add the URI parameter to example.env (Philip Chase)

### Changed
- Align Dockerfile with tt_engine work (Philip Chase)
- Use URI env var in load fake data (Philip Chase)


## [0.3.0] - 2020-04-24
### Added
- Set form_complete fields in PKY (Philip Chase)


## [0.2.0] - 2020-04-23
### Added
 - Verify the checksums of barcodes before save data to REDCap (Laurence James-Woodley)
 - Add scripts to load data nd test data for PKY (Philip Chase)
 - initialze consecutive negative swab results to zero (Philip Chase)

### Changed
 - Allow one docker image to run any of the scripts (Philip Chase)


## [0.1.0] - 2020-04-14
### Summary
 - First release of fr_covidata_engine
 - Supports swab results: pos, neg, and ina
 - Updates only blank results in survey project
 - Provides a log of upload results and non-loadable for swabs
 - Writes a custom email body using good/bad result count, project titles, PIDs, and URL to result project for address errors
 - Emails log file to configurable addressees


fr_etl.R
library(tidyverse)
library(dotenv)
library(REDCapR)

# NOTE: For production script the data will be read from pid 8270
# to get the research_encounter_id which will then be written to pid 8258
  
# set the timezone
Sys.setenv(TZ = Sys.getenv("TIME_ZONE"))

# read data from survey project
survey_project_read <- redcap_read_oneshot(redcap_uri = 'https://redcap.ctsi.ufl.edu/redcap/api/',
                               token = Sys.getenv("WRITE_TO"))$data %>%
  filter(!is.na(research_encounter_id)) %>%
  select(record_id, redcap_event_name, research_encounter_id, 
         covid_19_swab_result, igg_antibodies, igm_antibodies)

survey_swab_data <- survey_project_read %>%
  filter(is.na(covid_19_swab_result)) %>%
  select(record_id, redcap_event_name, research_encounter_id)

# read data from result upload project, pid 8258
upload_project_read <- redcap_read_oneshot(redcap_uri = 'https://redcap.ctsi.ufl.edu/redcap/api/',
                               token = Sys.getenv("READ_FROM"))$data

# make result upload file for swabs
swab_result <-
upload_project_read %>%
  filter(!is.na(record_id)) %>% 
  select(research_encounter_id = record_id, covid_19_swab_result) %>%
  filter(!is.na(covid_19_swab_result)) %>% 
  inner_join(survey_swab_data, by=c("research_encounter_id")) %>%
  mutate(covid_19_swab_result = case_when(
    str_detect(str_to_lower(covid_19_swab_result), "pos") ~ "1",
    str_detect(str_to_lower(covid_19_swab_result), "neg") ~ "0",
    TRUE ~ as.character(NA)
  )) %>%
  select(record_id, redcap_event_name, covid_19_swab_result) %>%
  arrange(record_id) 

# TODO:write data to survey project
# redcap_write_oneshot(records, 
#                      redcap_uri = 'https://redcap.ctsi.ufl.edu/redcap/api/',
#                      token = Sys.getenv("WRITE_TO"))



functions.R
verifyLuhn <- function(input, base_encoding = 16, factor = 2) {
  # https://en.wikipedia.org/wiki/Luhn_mod_N_algorithm
  # assumes input is hexadecimal
  # assumes input is in format 1-12345-01-X
  # <location_id>-<record_id>-<visit_id>-<check_digit>
  unpacked_input = strsplit(input, '-')
  check_digit = strtoi((unpacked_input[[1]][4]), base = base_encoding)
  # concatenate record_id and visit_id
  raw_number = paste(unpacked_input[[1]][2], unpacked_input[[1]][3], sep = '')
  
  # split the string, convert it to a list, convert digits to decimal
  digits = strtoi(unlist(strsplit(raw_number, "")), base = base_encoding)
  
  # reverse the string
  digits = digits[length(digits):1]
  
  # multiply every other digit by the proper factor
  to_mult = seq(1, length(digits), 2)
  digits[to_mult] = as.numeric(digits[to_mult]) * factor
  
  # sum the digits of the "addend" as expressed in base base_encoding
  addends = floor(digits / base_encoding) + (digits %% base_encoding)
  
  # returns a boolean
  return( sum(c(addends, strtoi(check_digit))) %% base_encoding == 0 )
}

load_fake_data_into_upload.R
library(tidyverse)
library(dotenv)
library(REDCapR)

# set the timezone
Sys.setenv(TZ = Sys.getenv("TIME_ZONE"))
Sys.getenv("INSTANCE")

records <- redcap_read_oneshot(redcap_uri = Sys.getenv("URI"),
                                       token = Sys.getenv("SURVEY_TOKEN"))$data %>% 
  filter(!is.na(research_encounter_id)) %>% 
  filter(is.na(covid_19_swab_result)) %>%
  select(research_encounter_id, covid_19_swab_result, igg_antibodies, igm_antibodies)

# note the number of records we have so we can make samples of the same size
n <- nrow(records)

# make some fake test results
results <- records %>%
  mutate(covid_19_swab_result = sample(c("Positive", "Negative", "negative", "positive"), n, replace = T)) %>%
  rename(record_id  = research_encounter_id) %>%
  mutate(igg_antibodies = sample(c("Positive", "Negative", "negative", "positive"), n, replace = T),
         igm_antibodies = sample(c("Positive", "Negative", "negative", "positive"), n, replace = T))

# Write data into the results project if it's safe to do so
if (Sys.getenv("INSTANCE") == "Development") {
  redcap_write_oneshot(results, 
                       redcap_uri = Sys.getenv("URI"),
                       token = Sys.getenv("RESULT_TOKEN"))
}




load_results_into_pky_projects.R
library(tidyverse)
library(lubridate)
library(dotenv)
library(REDCapR)
library(openxlsx)
library(sendmailR)

source("functions.R")

# echo details from the .env file we read
Sys.getenv("INSTANCE")
Sys.getenv("PROJECT")

# set the timezone
Sys.setenv(TZ = Sys.getenv("TIME_ZONE"))

script_run_time <- strftime(Sys.time(), format = "%Y%m%d_%H%M") 

# email credentials
email_server <- list(smtpServer = Sys.getenv("SMTP_SERVER"))
email_from <- Sys.getenv("EMAIL_FROM")
email_to <- unlist(strsplit(Sys.getenv("EMAIL_TO")," "))
email_cc <- unlist(strsplit(Sys.getenv("EMAIL_CC")," "))
email_subject <- paste(Sys.getenv("EMAIL_SUBJECT"), script_run_time)

# read data from survey project
survey_project_read <- redcap_read_oneshot(redcap_uri = 'https://redcap.ctsi.ufl.edu/redcap/api/',
                               token = Sys.getenv("SURVEY_TOKEN"))$data %>%
  filter(!is.na(research_encounter_id)) %>%
  mutate(test_date_and_time = strftime(test_date_and_time, "%Y-%m-%d %H:%M")) %>%
  mutate(research_encounter_id = as.character(research_encounter_id)) %>%
  mutate(covid_19_swab_result = as.numeric(covid_19_swab_result)) %>%
  mutate(subject_info_complete = 2) %>%
  mutate(appointment_complete = 2) %>%
  select(record_id,
         redcap_event_name,
         ce_firstname,
         ce_lastname,
         icf_age,
         patient_dob,
         ce_orgconsentdate,
         icf_email,
         qpk_phone,
         qpk_contact_name,
         subject_info_complete,
         research_encounter_id,
         test_date_and_time,
         appointment_complete,
         covid_19_swab_result)

# If data already in Survey got flushed from Serial (generally due to testing), load it into serial now
old_records_for_serial <- survey_project_read %>%
  filter(redcap_event_name == "baseline_arm_1") %>%
  filter(covid_19_swab_result == "1")

# write the new rows to serial if there were any
if(nrow(old_records_for_serial) > 0 ){
  redcap_write_oneshot(old_records_for_serial,
                       redcap_uri = 'https://redcap.ctsi.ufl.edu/redcap/api/',
                       token = Sys.getenv("SERIAL_TOKEN"))
}


# read data from the serial project...if there is any
serial_project_read_all <- redcap_read_oneshot(redcap_uri = 'https://redcap.ctsi.ufl.edu/redcap/api/',
                                           token = Sys.getenv("SERIAL_TOKEN"))
if(serial_project_read_all$success) {
  serial_project_read <- serial_project_read_all$data %>%
    filter(!is.na(research_encounter_id)) %>%
    mutate(research_encounter_id = as.character(research_encounter_id)) %>%
    select(record_id,
           redcap_event_name,
           research_encounter_id,
           test_date_and_time,
           consecutive_negative_swab_results,
           covid_19_swab_result)
  } else {
  serial_project_read <- tibble(record_id = character(),
                                redcap_event_name=character(),
                                research_encounter_id=character(),
                                test_date_and_time=ymd_hm(),
                                consecutive_negative_swab_results=numeric(),
                                covid_19_swab_result=logical())
}


# survey records without swab data
survey_swab_data <- survey_project_read %>%
  filter(is.na(covid_19_swab_result)) %>%
  select(record_id, redcap_event_name, research_encounter_id)

# serial records without swab data
serial_swab_data <- serial_project_read %>%
  filter(is.na(covid_19_swab_result)) %>%
  select(record_id, redcap_event_name, research_encounter_id)

# combined target rows
combined_read_data <- bind_rows(
  survey_project_read %>% select(record_id, redcap_event_name, research_encounter_id),
  serial_project_read %>% select(record_id, redcap_event_name, research_encounter_id)
  )


# read data from result upload project, (prod pid 8270)
result_project_read_all <- redcap_read_oneshot(redcap_uri = 'https://redcap.ctsi.ufl.edu/redcap/api/',
                               token = Sys.getenv("RESULT_TOKEN"))

if (result_project_read_all$success) {
  result_project_read <- result_project_read_all$data  %>%
    mutate(covid_19_swab_result = case_when(
      str_detect(str_to_lower(covid_19_swab_result), "pos") ~ "1",
      str_detect(str_to_lower(covid_19_swab_result), "neg") ~ "0",
      str_detect(str_to_lower(covid_19_swab_result), "ina") ~ "99",
      TRUE ~ covid_19_swab_result
    )) %>%
    rowwise() %>%
    mutate(verified_id = verifyLuhn(record_id))
} else {
  result_project_read <- tibble(record_id = character(),
                                covid_19_swab_result=numeric(),
                                igg_antibodies=numeric(),
                                igm_antibodies=numeric(),
                                verified_id=logical())
}


# Run these two tests on entire dataset so we'll know that the corrections have
# been made

# flag incorrect barcodes
result_id_with_bad_checksum <- result_project_read %>% 
  filter(!verified_id) %>% 
  mutate(reason_not_imported = 'bad checksum in barcode id') %>% 
  select(record_id, covid_19_swab_result, verified_id,
      reason_not_imported)

# on the low chance that the barcode passes checksum but does not match an id
# in the survey project
result_id_with_no_match_in_survey <- result_project_read %>% 
  filter(verified_id) %>% 
  anti_join(combined_read_data, by = c("record_id" = "research_encounter_id")) %>%
  mutate(reason_not_imported = 'no match in target project') %>% 
  select(record_id, covid_19_swab_result, verified_id, 
         reason_not_imported)

# make result upload file for swabs in the survey project
swab_result <- result_project_read %>% 
  filter(!is.na(record_id)) %>% 
  select(research_encounter_id = record_id, covid_19_swab_result, verified_id) %>%
  filter(!is.na(covid_19_swab_result)) %>%  
  # join to get records in survey project without swab results
  inner_join(survey_swab_data, by=c("research_encounter_id")) %>%
  select(record_id, redcap_event_name, covid_19_swab_result, verified_id) %>%
  filter(!(redcap_event_name != "baseline_arm_1" & covid_19_swab_result == 1)) %>%
  arrange(record_id)


# make result file for swab results that were scheduled in the Survey project 
# but need to be written to the serial project
swab_result_for_serial_spawned_from_survey <- result_project_read %>% 
  filter(!is.na(record_id)) %>% 
  select(research_encounter_id = record_id, covid_19_swab_result, verified_id) %>%
  filter(!is.na(covid_19_swab_result)) %>%  
  # join to get records in survey project without swab results
  inner_join(survey_swab_data, by=c("research_encounter_id")) %>%
  select(record_id, redcap_event_name, covid_19_swab_result, verified_id) %>%
  filter(redcap_event_name == "baseline_arm_1" & covid_19_swab_result == 1) %>%
  mutate(consecutive_negative_swab_results = 0) %>%
  arrange(record_id)

# seed new demographic data in serial for each new positive swab in the Survey project
new_records_for_serial <- swab_result_for_serial_spawned_from_survey %>% 
  filter(verified_id & covid_19_swab_result == 1) %>%
  select(record_id, redcap_event_name) %>%
  inner_join(survey_project_read) %>%
  mutate(test_date_and_time = strftime(test_date_and_time, "%Y-%m-%d %H:%M")) %>%
  select(-covid_19_swab_result)

# write the new rows to serial if there were any
if(nrow(new_records_for_serial) > 0 ){
  redcap_write_oneshot(new_records_for_serial,
                       redcap_uri = 'https://redcap.ctsi.ufl.edu/redcap/api/',
                       token = Sys.getenv("SERIAL_TOKEN"))
}

# make the dataset for the log file
new_records_for_serial_log_content <- new_records_for_serial %>%
  select(record_id,
         redcap_event_name,
         icf_age,
         ce_orgconsentdate,
         research_encounter_id,
         test_date_and_time)

# make result file for swab results that were scheduled in the Serial project 
# and need to be written to the serial project
swab_result_for_serial_spawned_from_serial <- result_project_read %>% 
  filter(!is.na(record_id)) %>% 
  select(research_encounter_id = record_id, covid_19_swab_result, verified_id) %>%
  filter(!is.na(covid_19_swab_result)) %>%  
  # join to get records in survey project without swab results
  inner_join(serial_swab_data, by=c("research_encounter_id")) %>%
  select(record_id, redcap_event_name, covid_19_swab_result, verified_id) %>%
  arrange(record_id)

swab_result_for_serial <- bind_rows(swab_result_for_serial_spawned_from_survey, 
                                swab_result_for_serial_spawned_from_serial)


all_swab_result <- bind_rows(swab_result, swab_result_for_serial)

# only send an email when there are new swab results
if (nrow(all_swab_result) > 0){

  # create folder to store output
  output_dir <- paste0("pky_covid19_import_log_", script_run_time)
  dir.create(output_dir, recursive = T)
  
  bad_swab_result <- all_swab_result %>% 
    select(-redcap_event_name) %>% 
    filter(!covid_19_swab_result %in% c("1","0", "99") & verified_id) %>% 
    mutate(reason_not_imported = "improper value for swab result") %>%   
    mutate_at(vars(record_id), as.character) %>% 
    bind_rows(result_id_with_bad_checksum) %>% 
    bind_rows(result_id_with_no_match_in_survey)

  # Write results to the survey project...
  swab_result_to_import <- swab_result %>% 
  filter(covid_19_swab_result %in% c("1","0", "99") & verified_id) %>% 
    select(-verified_id)
  
  # ...only write to redcap when there are legit records
  if(nrow(swab_result_to_import) > 0 ){
  redcap_write_oneshot(swab_result_to_import,
                       redcap_uri = 'https://redcap.ctsi.ufl.edu/redcap/api/',
                       token = Sys.getenv("SURVEY_TOKEN"))
  }

  
  # Write results to the serial project
  swab_result_serial_to_import <- swab_result_for_serial %>% 
    filter(covid_19_swab_result %in% c("1","0", "99") & verified_id) %>% 
    select(-verified_id)
  
  # ...only write to redcap when there are legit records
  if(nrow(swab_result_serial_to_import) > 0 ){
    redcap_write_oneshot(swab_result_serial_to_import,
                         redcap_uri = 'https://redcap.ctsi.ufl.edu/redcap/api/',
                         token = Sys.getenv("SERIAL_TOKEN"))
  }

  # make an Excel file with tabs for each slice of data
  all_output <- list("Swab Results Imported Scheduled" = swab_result_to_import,
                     "Swab Results Imported in Serial" = swab_result_serial_to_import,
                     "New records added to Serial" = new_records_for_serial_log_content,
                     "Swab Results Not Imported" = bad_swab_result)
  
  write.xlsx(all_output, 
             paste0(output_dir, "/swab_result_log_", script_run_time, ".xlsx"), 
             na = "")

  # Zip the reports generated
  zipfile_name <-  paste0(output_dir, ".zip")
  zip(zipfile_name, output_dir)
  
  # Write the email
  # Attach the zip file
  attachment_object <- mime_part(zipfile_name, zipfile_name)
  project_title <- Sys.getenv("SURVEY_PROJECT_TITLE")
  project_pid   <- Sys.getenv("SURVEY_PROJECT_PID")
  result_project_pid <- Sys.getenv("RESULT_PROJECT_PID")
  
  # Write the body of the email
  body <- paste0("The attached file(s) includes a log of all results that were uploaded",
               " to the REDCap project, ", project_title, " (PID ", project_pid ,")",
               "\n\nNumber of results loaded in Scheduled: ", nrow(swab_result_to_import),
               "\n\nNumber of results loaded in Serial: ", nrow(swab_result_serial_to_import),
               "\n\nNumber of records added to Serial: ", nrow(new_records_for_serial_log_content),
               "\nNumber of results not loaded: ", nrow(bad_swab_result),
               "\n\nIf there are records that were not loaded, then there were",
               " improper values in the swab result column or incorrect record_ids were used", 
               " Please review the Swab Results Not Imported tab in the attached log file",
               " then update these records at ",
               "https://redcap.ctsi.ufl.edu/redcap/redcap_v9.3.5/index.php?pid=", result_project_pid)
  
  body_with_attachment <- list(body, attachment_object)
  
  # send the email with the attached output file
  sendmail(from = email_from, to = email_to, cc = email_cc,
           subject = email_subject, msg = body_with_attachment,
           control = email_server)
  
  # uncomment to delete output once on tools4
  # unlink(zipfile_name)
  
} 


load_results_into_survey_project.R
library(tidyverse)
library(dotenv)
library(REDCapR)
library(openxlsx)
library(sendmailR)

source("functions.R")
# set the timezone
Sys.setenv(TZ = Sys.getenv("TIME_ZONE"))

# echo details from the .env file we read
Sys.getenv("INSTANCE")
Sys.getenv("PROJECT")

script_run_time <- strftime(Sys.time(), format = "%Y%m%d_%H%M") 

# email credentials
email_server <- list(smtpServer = Sys.getenv("SMTP_SERVER"))
email_from <- Sys.getenv("EMAIL_FROM")
email_to <- unlist(strsplit(Sys.getenv("EMAIL_TO")," "))
email_cc <- unlist(strsplit(Sys.getenv("EMAIL_CC")," "))
email_subject <- paste(Sys.getenv("EMAIL_SUBJECT"), script_run_time)

# NOTE: For production script the data will be read from pid 8270
# to get the research_encounter_id which will then be written to pid 8258

# read data from survey project (prod pid 8258)
survey_project_read <- redcap_read_oneshot(redcap_uri = 'https://redcap.ctsi.ufl.edu/redcap/api/',
                               token = Sys.getenv("SURVEY_TOKEN"))$data %>%
  filter(!is.na(research_encounter_id)) %>%
  select(record_id, redcap_event_name, research_encounter_id, 
         covid_19_swab_result, igg_antibodies, igm_antibodies)

# survey records without swab data
survey_swab_data <- survey_project_read %>%
  filter(is.na(covid_19_swab_result)) %>%
  select(record_id, redcap_event_name, research_encounter_id)

# read data from result upload project, (prod pid 8270)
result_project_read <- redcap_read_oneshot(redcap_uri = 'https://redcap.ctsi.ufl.edu/redcap/api/',
                               token = Sys.getenv("RESULT_TOKEN"))$data %>% 
  rowwise() %>% 
  mutate(verified_id = verifyLuhn(record_id))

# Run these two tests on entire dataset so we'll know that the corrections have
# been made

# flag incorrect barcodes
result_id_with_bad_checksum <- result_project_read %>% 
  filter(!verified_id) %>% 
  mutate(reason_not_imported = 'bad checksum in barcode id') %>% 
  select(record_id, covid_19_swab_result, verified_id,
      reason_not_imported)

# on the low chance that the barcode passes checksum but does not match an id
# in the survey project
result_id_with_no_match_in_survey <- result_project_read %>% 
  filter(verified_id) %>% 
  anti_join(survey_project_read, by = c("record_id" = "research_encounter_id")) %>% 
  mutate(reason_not_imported = 'no match in target project') %>% 
  select(record_id, covid_19_swab_result, verified_id, 
         reason_not_imported)

# make result upload file for swabs
swab_result <- result_project_read %>% 
  filter(!is.na(record_id)) %>% 
  select(research_encounter_id = record_id, covid_19_swab_result, verified_id) %>%
  filter(!is.na(covid_19_swab_result)) %>%  
  # join to get records in survey project without swab results
  inner_join(survey_swab_data, by=c("research_encounter_id")) %>%
  mutate(covid_19_swab_result = case_when(
    str_detect(str_to_lower(covid_19_swab_result), "pos") ~ "1",
    str_detect(str_to_lower(covid_19_swab_result), "neg") ~ "0",
    str_detect(str_to_lower(covid_19_swab_result), "ina") ~ "99",
    TRUE ~ covid_19_swab_result
  )) %>%
  select(record_id, redcap_event_name, covid_19_swab_result, verified_id) %>%
  arrange(record_id) 

# only send an email when there are new swab results
if (nrow(swab_result) > 0){

  # create folder to store output
  output_dir <- paste0("fr_covid19_import_log_", script_run_time)
  dir.create(output_dir, recursive = T)
  
  # write data to survey project
  swab_result_to_import <- swab_result %>% 
  filter(covid_19_swab_result %in% c("1","0", "99") & verified_id) %>% 
    select(-verified_id)
  
  bad_swab_result <- swab_result %>% 
    select(-redcap_event_name) %>% 
    filter(!covid_19_swab_result %in% c("1","0", "99") & verified_id) %>% 
    mutate(reason_not_imported = "improper value for swab result") %>%   
    mutate_at(vars(record_id), as.character) %>% 
    bind_rows(result_id_with_bad_checksum) %>% 
    bind_rows(result_id_with_no_match_in_survey)
  
  # only write to redcap when there are legit records
  if(nrow(swab_result_to_import) > 0 ){
  redcap_write_oneshot(swab_result_to_import,
                       redcap_uri = 'https://redcap.ctsi.ufl.edu/redcap/api/',
                       token = Sys.getenv("SURVEY_TOKEN"))
  }
  
  all_output <- list("Swab Results Imported" = swab_result_to_import,
                     "Swab Results Not Imported" = bad_swab_result)
  
  write.xlsx(all_output, 
             paste0(output_dir, "/swab_result_log_", script_run_time, ".xlsx"), 
             na = "")
  
  # Zip the reports generated
  zipfile_name <-  paste0(output_dir, ".zip")
  zip(zipfile_name, output_dir)
  
  # attach the zip file and email it
  attachment_object <- mime_part(zipfile_name, zipfile_name)
  project_title <- Sys.getenv("SURVEY_PROJECT_TITLE")
  project_pid   <- Sys.getenv("SURVEY_PROJECT_PID")
  result_project_pid <- Sys.getenv("RESULT_PROJECT_PID")
  
  body <- paste0("The attached file(s) includes a log of all results that were uploaded",
               " to the REDCap project, ", project_title, " (PID ", project_pid ,")",
               "\n\nNumber of records uploaded: ", nrow(swab_result_to_import),
               "\nNumber of records not uploaded: ", nrow(bad_swab_result),
               "\n\nIf there are records that were not uploaded, then there were",
               " improper values in the swab result column or incorrect record_ids were used", 
               " Please review the Swab Results Not Imported tab in the attached log file",
               " then update these records at ",
               "https://redcap.ctsi.ufl.edu/redcap/redcap_v9.3.5/index.php?pid=", result_project_pid)
  
  body_with_attachment <- list(body, attachment_object)
  
  # send the email with the attached output file
  sendmail(from = email_from, to = email_to, cc = email_cc,
           subject = email_subject, msg = body_with_attachment,
           control = email_server)
  
  # uncomment to delete output once on tools4
  # unlink(zipfile_name)
  
} 


load_fr_igg_results_into_survey_project.R
library(tidyverse)
library(dotenv)
library(REDCapR)
library(openxlsx)
library(sendmailR)

source("functions.R")
# set the timezone
Sys.setenv(TZ = Sys.getenv("TIME_ZONE"))

# echo details from the .env file we read
Sys.getenv("INSTANCE")
Sys.getenv("PROJECT")

script_run_time <- strftime(Sys.time(), format = "%Y%m%d_%H%M") 

# email credentials
email_server <- list(smtpServer = Sys.getenv("SMTP_SERVER"))
email_from <- Sys.getenv("EMAIL_FROM")
email_to <- unlist(strsplit(Sys.getenv("EMAIL_TO")," "))
email_cc <- unlist(strsplit(Sys.getenv("EMAIL_CC")," "))
email_subject <- paste(Sys.getenv("EMAIL_SUBJECT"), script_run_time)

# NOTE: For production script the data will be read from pid 8270
# to get the research_encounter_id which will then be written to pid 8258

# read data from survey project (prod pid 8258)
survey_project_read <- redcap_read_oneshot(redcap_uri = Sys.getenv("URI"),
                               token = Sys.getenv("SURVEY_TOKEN"))$data %>%
  filter(!is.na(research_encounter_id)) %>%
  select(record_id, redcap_event_name, research_encounter_id, 
         covid_19_swab_result, igg_antibodies, igm_antibodies)

# survey records without a lab result
survey_lab_data <- survey_project_read %>%
  filter(is.na(igg_antibodies)) %>%
  select(record_id, redcap_event_name, research_encounter_id)

# read data from result upload project, (prod pid 8270)
result_project_read <- redcap_read_oneshot(redcap_uri = Sys.getenv("URI"),
                               token = Sys.getenv("RESULT_TOKEN"))$data %>% 
  rowwise() %>% 
  mutate(verified_id = verifyLuhn(record_id)) %>%
  mutate(igg_antibodies = as.character(igg_antibodies)) %>%
  mutate(igm_antibodies = as.character(igm_antibodies))


# Run these two tests on entire dataset so we'll know that the corrections have
# been made

# flag incorrect barcodes
result_id_with_bad_checksum <- result_project_read %>% 
  filter(!verified_id) %>% 
  mutate(reason_not_imported = 'bad checksum in barcode id') %>% 
  select(record_id, igg_antibodies, verified_id,
      reason_not_imported)

# on the low chance that the barcode passes checksum but does not match an id
# in the survey project
result_id_with_no_match_in_survey <- result_project_read %>% 
  filter(verified_id) %>% 
  anti_join(survey_project_read, by = c("record_id" = "research_encounter_id")) %>% 
  mutate(reason_not_imported = 'no match in target project') %>% 
  select(record_id, igg_antibodies, verified_id, 
         reason_not_imported)

# make result upload file for this test
lab_result <- result_project_read %>% 
  filter(!is.na(record_id)) %>% 
  select(research_encounter_id = record_id, igg_antibodies, verified_id) %>%
  filter(!is.na(igg_antibodies)) %>%  
  # join to get records in survey project without swab results
  inner_join(survey_lab_data, by=c("research_encounter_id")) %>%
  mutate(igg_antibodies = case_when(
    str_detect(str_to_lower(igg_antibodies), "pos") ~ "1",
    str_detect(str_to_lower(igg_antibodies), "neg") ~ "0",
    str_detect(str_to_lower(igg_antibodies), "indeterminate") ~ "2",
    str_detect(str_to_lower(igg_antibodies), "inadequate") ~ "99",
    TRUE ~ igg_antibodies
  )) %>%
  select(record_id, redcap_event_name, igg_antibodies, verified_id) %>%
  arrange(record_id) 

# only send an email when there are new swab results
if (nrow(lab_result) > 0){

  # create folder to store output
  output_dir <- paste0("fr_covid19_import_log_", script_run_time)
  dir.create(output_dir, recursive = T)
  
  # write data to survey project
  lab_result_to_import <- lab_result %>% 
  filter(igg_antibodies %in% c("1","0","2","99") & verified_id) %>% 
    select(-verified_id)
  
  bad_lab_result <- lab_result %>% 
    select(-redcap_event_name) %>% 
    filter(!igg_antibodies %in% c("1","0","2","99") & verified_id) %>% 
    mutate(reason_not_imported = "improper value for result") %>%   
    mutate_at(vars(record_id), as.character) %>% 
    bind_rows(result_id_with_bad_checksum) %>% 
    bind_rows(result_id_with_no_match_in_survey)
  
  # only write to redcap when there are legit records
  if(nrow(lab_result_to_import) > 0 ){
  redcap_write_oneshot(lab_result_to_import,
                       redcap_uri = Sys.getenv("URI"),
                       token = Sys.getenv("SURVEY_TOKEN"))
  }
  
  all_output <- list("Results Imported" = lab_result_to_import,
                     "Results Not Imported" = bad_lab_result)
  
  write.xlsx(all_output, 
             paste0(output_dir, "/lab_result_log_", script_run_time, ".xlsx"), 
             na = "")
  
  # Zip the reports generated
  zipfile_name <-  paste0(output_dir, ".zip")
  zip(zipfile_name, output_dir)
  
  # attach the zip file and email it
  attachment_object <- mime_part(zipfile_name, zipfile_name)
  project_title <- Sys.getenv("SURVEY_PROJECT_TITLE")
  project_pid   <- Sys.getenv("SURVEY_PROJECT_PID")
  result_project_pid <- Sys.getenv("RESULT_PROJECT_PID")
  
  body <- paste0("The attached file(s) includes a log of all results that were uploaded",
               " to the REDCap project, ", project_title, " (PID ", project_pid ,")",
               "\n\nNumber of records uploaded: ", nrow(lab_result_to_import),
               "\nNumber of records not uploaded: ", nrow(bad_lab_result),
               "\n\nIf there are records that were not uploaded, then there were",
               " improper values in the result column or incorrect record_ids were used", 
               " Please review the Results Not Imported tab in the attached log file",
               " then update these records at ",
               "https://redcap.ctsi.ufl.edu/redcap/redcap_v9.3.5/index.php?pid=", result_project_pid)
  
  body_with_attachment <- list(body, attachment_object)
  
  # send the email with the attached output file
  sendmail(from = email_from, to = email_to, cc = email_cc,
           subject = email_subject, msg = body_with_attachment,
           control = email_server)
  
  # uncomment to delete output once on tools4
  # unlink(zipfile_name)
  
} 
